### ISSUE TITLE
Slow Performance of scanner

### ISSUE THREAD

**Issue by Jazz9** [minio/minio#18679]
* reopened due to my poor description of the fault on #18677 

I am trying to scan a minio cluster to ensure data is valid.

![image](https://github.com/minio/minio/assets/39378577/3e69db42-44e3-44e4-8372-946a48f46606)

The individual drive scanning speed is good. The drive that are scanning are 140 IOPS / 2-3 queue depth, 40 ms read avg.

The rest of the cluster is not busy. How do we increase the number of drives scanning?

## Expected Behavior
On fastest setting - most of the cluster should be scanning.

## Current Behavior
Drives scanning starts at 12 and over 24 hours drops to 1-2 drives.

## Possible Solution
Increase the drives currently scanning

Steps to Reproduce (for bugs)
Reformat server with HDD drives - wait 24h. Have a cluster with 200+TB of ~1M files.

Context
Likelihood of data loss increases with the excessive time to rebuild a node.

Your Environment
6 x ubuntu 22 servers with 12 x enterprise SATA 18TB drives, 2 x 10gbit, 192GB ram.
Version RELEASE.2023-12-14T18-51-57Z

**Comment by Jazz9** [comment]
This is what the scanner info looks like when you restart the minio process 

![image](https://github.com/minio/minio/assets/39378577/aad554a5-61e3-4d69-b681-36e5db0f9c1f)

**Comment by harshavardhana** [comment]
Reopening the issue won't get you anywhere @Jazz9 yes that is normal, you need to let the scanner cycles complete.

If you restart them in between it will have no way to visit the new set of folders in the next run it seems like you are somehow under the impression that it's some kind of blind scan on the namespace and there are some nuances to it. 

When it says there are only two drives being scanned it means that it has already done the rest and it is waiting for the last remaining ones to finish the cycle. 

Until the cycle fully completes it doesn't schedule another run on another set of drives. We are planning to change this in couple of PRs which is not merged yet.

You cannot arbitrarily speed it up, that is not how it works. It is always going to be a function of the slowest component on the system which is your drives.

To manually trigger if any missed heals during active run during drive replacements.

```
mc admin heal -r --remove alias/
```

Scanner is not meant for rebuilding your drives, it gets to them in 512th cycle and later.

Your individual cycles are slow, mostly due to a badly sized setup for wrong workload to begin with using small files with HDDs.

**Comment by Jazz9** [comment]
Thanks for the explanation of the issue. 

Is there a way to increase the number of drives in a cycle?

**Comment by harshavardhana** [comment]
@Jazz9 That is automatic. It's based on the number of buckets you have to the number of drive relationships per erasure set.

The more erasure sets, the better it is for the scanner; if you have a wider stripe, the work becomes limited, which may also be your problem, along with slower disks for the narrow blocks we have to read for small objects.

IOPS are limited on HDDs, so if you have smaller objects, the work that needs to be done by the scanner is longer; HDDs being already starved of IOPs, you are limited by the speed for metadata reads and lookups.

I want to know more about the original problem where you indicated that healing was finished but not complete @Jazz9 ? Can you provide more details about this problem? Can you open a separate GitHub issue for that? We want to know under what conditions that could have happened.

Thank you

**Comment by klauspost** [comment]
The 39ms to open and read a single `xl.meta` is your problem. We do not provide performance optimization as a community service, for that look into a [subscription](https://min.io/pricing).

**Comment by Jazz9** [comment]
> The 39ms to open and read a single `xl.meta` is your problem. We do not provide performance optimization as a community service, for that look into a [subscription](https://min.io/pricing).

The issue is that it isn't doing enough in parallel. I understand that it is stressing 1-2 drives in the 70 disk cluster to 39ms but that is far from a great outcome for performance.

**Comment by klauspost** [comment]
@Jazz9 It is what is is. We may or may not change how it operates in the future - but I think we have reached the limit to what is useful to discuss here.
